### ISSUE TITLE
Can we eliminate some CSS custom properties in favor of standard CSS properties?

### ISSUE THREAD

**Issue by LeaVerou** [shoelace-style/webawesome#259]
We currently have a lot of [component-specific CSS properties](https://webawesome-git-next-font-awesome.vercel.app/docs/components/reference#all-css-custom-properties) that basically mirror native CSS properties.

From a quick skim, these are the obvious ones:

- `--background-color` (19 components)
- `--border-radius` (15 components)
- `--border-color` (14 components)
- `--border-style` (14 components) 	
- `--border-width`  (14 components)
- `--box-shadow` (13 components)
- `--width` (3 components)
- `--height` (2 components)
- `--aspect-ratio` (2 components)
- `--padding`  (2 components)
- `--max-width` (`<wa-tooltip>`)

<details>
<summary>And a few more that may be more controversial:</summary>

- `--size` for `<wa-avatar>` and `<wa-progress-ring>`  (for  `<wa-drawer>` it seems to serve a distinct purpose). If we use `aspect-ratio: 1`, people can just set either `width` or `height` and it would work fine.
- `--color`, used in `<wa-divider>` and `<wa-skeleton>`. Neither of these components includes text, so we can just make `color` work by setting `background-color: currentColor;`.
- `--slide-gap` (`<wa-carousel>`): Could we use `gap`? I suspect this would be a tough one, but including it in case.
`--primary-color` and `--primary-opacity`	(`<wa-icon>`): Instead of having separate properties for the primary and secondary colors, we could just use `color` and `opacity` for the primary color (which is also _the_ color for non duotone icons) and only have custom properties for the secondary color/opacity.
- `--menu-width`, `--main-width`, `--aside-width`, `--banner-height`,  `--header-height`, `--subheader-height` (`<wa-page>`): Do we need these? Aren't there always author-provided elements there? If not, feel free to ignore this.
- `--symbol-spacing` (`<wa-rating>`): Can we use `gap`?
- `--symbol-color` (`<wa-rating>`): Can we just use `color`?
- `--symbol-size` (`<wa-rating>`): Can we just use `font-size` or `height`?
- `--min` and `--max` (`<wa-split-panel>`): Can we just use `min-width`/`max-width`/`min-height`/`max-height`? Sure, you lose the fact that the current `--min` and `--max` adjusts to the orientation, but how commonly do people switch between orientations anyway? I suspect the gain in understandability and knowledge portability of using the standard attributes would be worth it.
</details>

In many cases this is a necessary evil:
- We may want to use these in multiple places in the component, often in different properties than their name (e.g. the `pulse` effect in badges defaults to the background color)
- We may want to transform them somehow (e.g. mixing a color)

But the more WA components can work with regular CSS, the better, as authors can reuse their existing knowledge rather than having to learn new API surface. It's not about difficulty — it's trivial to use a custom property instead of a standard one if you know about it. It's about:
1. **Friction**: unless you already know about this, it usually goes like this: you try the standard property, try to debug why it doesn't work, eventually look up the docs, find out you need to use a different property. 
2. **Interoperability**: Custom properties introduce a coupling with WA. Third-party tools and libraries do not know about our custom properties, but if they are setting regular properties things Just Work™.

**Comment by claviska** [comment]
Some of these seem easy enough to swap out. Others, maybe not. Dialog, for example, uses `--width` to apply to the dialog body _within_ the shadow root. Maybe it's possible, but I'm not sure how to pass width from the host element through the other elements or if I would even expect that as a user. Perhaps a more explicit name would be better: `--dialog-width`

**Comment by LeaVerou** [comment]
> Some of these seem easy enough to swap out. Others, maybe not. 

Yup, I was hoping to get exactly those insights from you!

> Dialog, for example, uses `--width` to apply to the dialog body _within_ the shadow root. Maybe it's possible, but I'm not sure how to pass width from the host element through the other elements or if I would even expect that as a user. Perhaps a more explicit name would be better: `--dialog-width`

Could you explain this a bit more? E.g. with an example? Is the host not rendered at all?

**Comment by claviska** [comment]
The structure of a dialog looks like this:

![CleanShot 2024-12-02 at 21 44 10@2x](https://github.com/user-attachments/assets/e86ab512-a7f0-4aee-ba9d-1505930f2f13)

The highlighted part ("panel") is the one that gets the width, in this case. It might be possible to pass width through without affecting the overlay, though. If we think it's worth it, that is. 🤔

**Comment by LeaVerou** [comment]
Why don’t we use `<dialog>` in there? It’s supported everywhere now and we’d get the overlay for free. Plus, real modals.

**Comment by claviska** [comment]
That's a Shoelace screenshot. Web Awesome uses `<dialog>` now.

**Comment by LeaVerou** [comment]
Oh great. Then where is the indirection? Shouldn't setting `width` on the `<wa-dialog>` itself work?

**Comment by claviska** [comment]
Generally speaking, with a property such as `width` being applied to the host element, I can't reliably pass that style through to the appropriate element(s) in a component. (It might be possible in dialog, it may or may not require structural changes, I'm not sure about this particular example.)

I'm just saying if we commit to that pattern, which would indeed be nice for certain properties, we may not be able to apply it consistently to all elements we want. I guess, for those cases, custom properties are a viable option.

**Comment by LeaVerou** [comment]
> Generally speaking, with a property such as `width` being applied to the host element, I can't reliably pass that style through to the appropriate element(s) in a component. (It might be possible in dialog, it may or may not require structural changes, I'm not sure about this particular example.)

Yeah, I get it in general. Though in this case it seems setting the width of the host would naturally propagate to the element itself simply because that's how CSS layout works for extrinsic sizing (which is what is used here).

> I'm just saying if we commit to that pattern, which would indeed be nice for certain properties, we may not be able to apply it consistently to all elements we want. I guess, for those cases, custom properties are a viable option.

I think it's fine to use custom properties where we definitely need custom properties and standard properties where we can. The mix doesn't introduce an inconsistency because we _already_ support standard properties for many things.

**Comment by LeaVerou** [comment]
The start has now been made with `<wa-callout>` so I've been thinking: one issue with removing `base` is that styles on `:host` have a much lower cascade order than anything styling the component from the outside. This is _usually_ not a problem because most selectors are more specific, but there are three properties where it could pose an issue, because people often do things like:

```css
* {
	padding: 0;
	margin: 0;
	box-sizing: border-box;
}
```

Most of our components don't set margins, and for `box-sizing` we already set it to `border-box` with `!important`, so this only leaves `padding`, which we may want to continue using custom properties for (with `padding: var(--padding) !important`). 🤔

**Comment by LeaVerou** [comment]
I took a look today and it seems that fixing #207 largely eliminates this issue as well. Most of the remaining properties just need renaming to make their purpose more clear.

E.g. looking at these properties on `<wa-input>`, they exist to apply styles on the actual input container. This is a legit use of custom properties, since the component consists of labels, hints and other elements too. It seems that they just need a different name to make their purpose more clear and prevent [confusion like this customer's](https://github.com/shoelace-style/webawesome-alpha/discussions/151). For example, `--control-background-color` instead of just `--background-color`. Ideally, naming would follow the part name, so that once people know what the part is called and the general naming scheme, they can predict what the properties are called and vice versa.
